tsl/src/continuous_aggs/invalidation_threshold.c

```diff
@@ -66,47 +67,67 @@
 
 typedef struct InvalidationThresholdData
 {
-	int64 threshold;
-	bool was_updated;
+	const ContinuousAgg *cagg;
+	const InternalTimeRange *refresh_window;
+	int64 computed_invalidation_threshold;
 } InvalidationThresholdData;
 
-static ScanTupleResult scan_update_invalidation_threshold(TupleInfo *ti, void *data);
-static ScanTupleResult invalidation_threshold_tuple_found(TupleInfo *ti, void *data);
-
 static ScanTupleResult
-scan_update_invalidation_threshold(TupleInfo *ti, void *data)
+invalidation_threshold_scan_update(TupleInfo *ti, void *const data)
 {
-	InvalidationThresholdData *invthresh = data;
-	bool should_free;
-	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
-	Form_continuous_aggs_invalidation_threshold form =
-		(Form_continuous_aggs_invalidation_threshold) GETSTRUCT(tuple);
+	InvalidationThresholdData *invthresh = (InvalidationThresholdData *) data;
+
+	if (ti->lockresult != TM_Ok)
+		return SCAN_CONTINUE;
+
+	bool isnull;
```

Comment:
I was trying to find in the code how we handle NULL values, because when we insert the initial threshold, the code seems to imply we are inserting a NULL value (see call of `invalidation_threshold_insert()`)

I spent some time looking for NULL handling but later realized that NULL is actually never inserted because it is turned into a MIN value. But this wasn't obvious. I made another comment on this below.

Reply:
Actually with this new implementation we'll never end up with a null value... actually here I can add a `Ensure` call because if a null value happen then seems there's something really wrong with the metadata

Reply:
Have a look at the `Ensure` I've added bellow